plugins/action/junos_facts.py

```diff
@@ -0,0 +1,151 @@
+        if not (
+            hasattr(self._connection, "socket_path")
+            and self._connection.socket_path is not None
+        ):
+            raise AnsibleConnectionError(
+                "netconf connection to remote host in not active"
+            )
+        if self._result.get("failed"):
+            return self._result
+
+        module = AnsibleModule(
+            argument_spec=argument_spec, supports_check_mode=True
+        )
+        (module._socket_path) = self._connection.socket_path
+        warnings = []
+        if task_vars["gather_subset"] == "!config":
+            warnings.append(
+                "default value for `gather_subset` will be changed to `min` from `!config` v2.11 onwards"
+            )
+        ansible_facts = {}
+        if task_vars.get("available_network_resources"):
+            ansible_facts["available_network_resources"] = sorted(
+                FACT_RESOURCE_SUBSETS.keys()
+            )
+
+        ansible_network_resources = {}
+
+        restorun_subsets = Facts(module).get_restorun()
+        result = Facts(module).get_legacy_facts()
+        additional_facts, additional_warnings = result
+        ansible_facts.update(additional_facts)
+        warnings.extend(additional_warnings)
+
+        if restorun_subsets:
+            for rmodule in restorun_subsets:
+                name = rmodule
+                rmodule = (
+                    self._task.collections[0]
+                    + "."
+                    + self._task.collections[0].split(".")[1]
+                    + "_"
+                    + rmodule
+                )
+                if not self._shared_loader_obj.module_loader.has_plugin(
+                    rmodule
+                ):
+                    result.update(
+                        {
+                            "failed": True,
+                            "msg": "Could not find %s module." % rmodule,
+                        }
+                    )
+                else:
+                    self._display.vvvv(
+                        "Running %s module to fetch data from remote host"
+                        % rmodule
+                    )
+                    result = self._execute_module(
+                        module_name=rmodule,
+                        module_args={"state": "gathered"},
+                        task_vars=task_vars,
+                        wrap_async=self._task.async_val,
+                    )
+                    if result.get("failed"):
+                        return result
+
+                    if result.get("gathered"):
+                        ansible_network_resources[name] = result["gathered"]
+
+        ansible_facts.update(
+            {"ansible_network_resources": ansible_network_resources}
+        )
+        ansible_facts[
+            "ansible_net_gather_network_resources"
+        ] = self._task.args.get("gather_network_resources")
+        ansible_facts["gather_subset"] = module.params.get("gather_subset")
```

Comment:
All the lines from creating the `AnsibleModule`  instance can be abstracted out in a base class method (maybe in netcommon collection) and can be reused by other platform facts action plugin to reduce code duplication.

Doing this will require a major release of the collection as it will require netcommon collection dependent version upgrade.